### ISSUE TITLE
[Feature] Save all shsh blobs that are currently being signed.

### ISSUE THREAD

**Issue by simon66** [tihmstar/tsschecker#37]
tsschecker_windows.exe -d AppleTV5,3 -s -e FFFFFFFFFFFFFF -i 10.0.1

Also, of course "FFFFFFFFFFFFFF" is not my ECID :)

```
Version: af9b9f2998bdeca7ece0829c2e718ae747745750 - 170
[TSSC] manually specified ecid to use, parsed "nothing to see here"
[TSSC] opening firmware.json
[JSON] counting elements
[JSON] parsing elements
[TSSC] using cached Buildmanifest for AppleTV5,3_10.0.1
[TSSC] opening bbgcid.json
[JSON] counting elements
[JSON] parsing elements
[Error] [TSSC] parsing bbgcid.json failed!
[WARNING] [TSSR] WARNING: there was an error getting BasebandGoldCertID, continuing without requesting Baseband ticket
[TSSR] Request URL set to https://gs.apple.com/TSS/controller?action=2
[TSSR] Sending TSS request attempt 1... failure

iOS 10.0.1 for device AppleTV5,3 IS NOT being signed!
```

As you can see, tsschecker is complaining that tvOS 10.0.1 is not being signed even though it is (https://ipsw.me/)

Not really sure why this is giving me an error :S

**Comment by derobi** [comment]
can you show me the line of code you ran without your ECID# please

**Comment by simon66** [comment]
```
tsschecker_windows.exe -d AppleTV5,3 -s -i 10.0.1
Version: af9b9f2998bdeca7ece0829c2e718ae747745750 - 170
[TSSC] opening firmware.json
[JSON] counting elements
[JSON] parsing elements
[TSSC] using cached Buildmanifest for AppleTV5,3_10.0.1
[TSSC] opening bbgcid.json
[JSON] counting elements
[JSON] parsing elements
[Error] [TSSC] parsing bbgcid.json failed!
[WARNING] [TSSR] WARNING: there was an error getting BasebandGoldCertID, continuing without requesting Baseband ticket
[TSSR] Request URL set to https://gs.apple.com/TSS/controller?action=2
[TSSR] Sending TSS request attempt 1... failure

iOS 10.0.1 for device AppleTV5,3 IS NOT being signed!
```
Same error.

**Comment by derobi** [comment]
add --nocache to the end of your ipsw for example 10.0.1 --nocache

**Comment by simon66** [comment]
```
tsschecker_windows.exe -d AppleTV5,3 -s -i 10.0.1 --nocache
Version: af9b9f2998bdeca7ece0829c2e718ae747745750 - 170
[TSSC] opening firmware.json
[DOWN] downloading file https://api.ipsw.me/v2.1/firmwares.json/condensed
[JSON] counting elements
[JSON] parsing elements
[TSSC] got firmwareurl for iOS 10.0.1 build 14U100
[TSSC] opening Buildmanifest for AppleTV5,3_10.0.1
[LFZP] downloading BuildManifest.plist from http://appldnld.apple.com/tvos10.0.1
/031-91431-20161109-3F9F616A-A770-11E6-BD96-A8C6F9858D7F/AppleTV5,3_10.0.1_14U100_Restore.ipsw

←[A←[J100 [===================================================================================================>]
←[A←[J037 [====================================>]
←[A←[J100 [===================================================================================================>]
[TSSC] opening bbgcid.json
[DOWN] downloading file http://api.tihmstar.net/bbgcid?condensed=1
[JSON] counting elements
[JSON] parsing elements
[Error] [TSSC] ERROR: device "AppleTV5,3" is not in bbgcid.json, which means it's BasebandGoldCertID isn't documented yet.
If you own such a device please consider contacting @tihmstar (tihmstar@gmail.com) to get instructions how to contribute to this project.
[WARNING] [TSSR] WARNING: there was an error getting BasebandGoldCertID, continuing without requesting Baseband ticket
[TSSR] Request URL set to https://gs.apple.com/TSS/controller?action=2
[TSSR] Sending TSS request attempt 1... failure

iOS 10.0.1 for device AppleTV5,3 IS NOT being signed!
```

**Comment by tihmstar** [comment]
```tihms-MBP:~ tihmstar$ tsschecker -d AppleTV5,3 -l
Version: a29fb9f9495f3a93571457c43edab530c3f85f07 - 180
[TSSC] opening firmware.json
[JSON] counting elements
[JSON] parsing elements
[TSSC] selecting latest iOS: 10.1
[TSSC] using cached Buildmanifest for AppleTV5,3_10.1
[WARNING] [TSSC] A BasebandGoldCertID is not required for AppleTV5,3
[TSSR] LOG: device AppleTV5,3 doesn't need a Baseband ticket, continuing without requesting a Baseband ticket
[TSSR] Request URL set to https://gs.apple.com/TSS/controller?action=2
[TSSR] Sending TSS request attempt 1... success

iOS 10.1 for device AppleTV5,3 IS being signed!
tihms-MBP:~ tihmstar$ 
```

Have you considered the fact that 10.0.1 is simply not signed at the moment?

**Comment by simon66** [comment]
> Have you considered the fact that 10.0.1 is simply not signed at the moment?
Do you actually think I didn't check that first? If you check the link in the first post you will see that ios 10.0.1 is still being signed :+1: 

iOS 10.0.1 is still being signed (I just re-updated my AppleTV to 10.0.1).
For some reason, when I run the "-i 10.0.1" flag, tsschecker doesn't check the latest build of the iOS. So what's happening is, iOS 10.0.1 has 2 builds (14U100 and 14U71). The latter being the most up to date build. While both builds are for the same iOS version, 14U100 is not being signed but 14U71 is.
tsschecker should check to see if all builds are still being signed (instead of checking just one).

This sure sounds like a bug to me :)

```
tsschecker_windows.exe -d AppleTV5,3 -s -e **** -i 10.0.1 --buildid 14U71
[TSSC] manually specified ecid to use, parsed "****" to dec:**** hex:****
[TSSC] opening firmware.json
[JSON] counting elements
[JSON] parsing elements
[TSSC] got firmwareurl for iOS 10.0.1 build 14U71
[TSSC] opening Buildmanifest for AppleTV5,3_14U71
[LFZP] downloading BuildManifest.plist from http://appldnld.apple.com/tvos10.0.1/031-59447-20161024-AA217A0A-9483-11E6-A69F-635E80A31755/AppleTV5,3_10.0.1_14U71_Restore.ipsw

←[A←[J100 [===================================================================================================>]
←[A←[J037 [====================================>
                               ]
←[A←[J100 [===================================================================================================>]
[TSSC] opening bbgcid.json
[JSON] counting elements
[JSON] parsing elements
[Error] [TSSC] ERROR: device "AppleTV5,3" is not in bbgcid.json, which means it's BasebandGoldCertID isn't documented yet.
If you own such a device please consider contacting @tihmstar (tihmstar@gmail.com) to get instructions how to contribute to this project.
[WARNING] [TSSR] WARNING: there was an error getting BasebandGoldCertID, continuing without requesting Baseband ticket
[TSSR] Request URL set to https://gs.apple.com/TSS/controller?action=2
[TSSR] Sending TSS request attempt 1... success
Saved shsh blobs!

**Build 14U71 for device AppleTV5,3 IS being signed!**
```

**Comment by tihmstar** [comment]
Why haven't you noted that iOS 10.0.1 has two builds and that one of them isn't signed before?
Would have made things easier as it now actually points to the issue you're refearing to. 
People open lots of issues on this project with stuff like "crashes on windows". I've already seen that people report "issues" with tsschecker because a version they want a ticket for wasn't signed anymore. 

Now to your actual issue. 
Since there are apperently 2 builds with the same version number this is indeed an interesting problem. 
I wouldn't say it's a bug, because tsschecker does what it's supposed to do, which is:
-get a list of devices/firmwares from api.ipsw.me
-find the correct ipsw
-download the manifest
-send a tss request

Since you know which build you want to check signing status (or save shsh) you can already do this with --buildid. 
I see that it could be cool to an option like --agressive or --check-all-builds or sth to tsschecker to check all builds for a specific iOS version. One could argue whether this should be default behavior or optional via parameter, but in my opinion this is definetly more of an enhancement/feature request than a bug. 
Though it's definetly something i'll look into :)

**Comment by simon66** [comment]
To be honest, when you closed the issue I wanted to investigate more into it, and thats when I found out about the 2 builds haha.
I know, I know, it would've made things much easier haha.

I feel like by default tsschecker should download every signed iOS version the user specify (If they use the -l param, tsschecker should check into all builds for the latest iOS). 

I'm looking into this now. Would you mind if I implemented it 👍 ?

**Comment by tihmstar** [comment]
To be honest i'd rather implement myself since quite a few things need to be adjusted.
I'll try to implement it this week maybe :)

**Comment by simon66** [comment]
Oh no worries then :)

I've implemented the fix on my local copy (Wanted to get the shsh for all my devices before they stopped signing them) .

I'll wait for the official fix :)

**Comment by tihmstar** [comment]
added this in cb5d1308ad274ef02c42b4cb0cb67c64c1bfc553
please check whether this works as intended.
Also please notify me if this commit caused any other bugs :)
